### PR TITLE
add layer2 sdk

### PR DESCRIPTION
**Pull request by blockchain-develop**

### CODE REVIEW

File: client/client.go [hunk 1]
```diff
@@ -350,6 +387,51 @@ func (this *ClientMgr) PreExecTransaction(mutTx *types.MutableTransaction) (*sdk
 	return preResult, nil
 }
 
+func (this *ClientMgr) VerifyStoreProof(key []byte, value []byte, proof []byte, stateRoot []byte) (bool, error) {
```
Comment (laizy) [inline client/client.go:390]: move to Layer2ClientMgr

File: client/client.go [hunk 2]
```diff
@@ -350,6 +387,51 @@ func (this *ClientMgr) PreExecTransaction(mutTx *types.MutableTransaction) (*sdk
...
+		return false, err
+	}
+	err = proof_iavl.VerifyItem(key, value)
+	if err != nil {
+		return false, err
+	}
+	return true, nil
+}
+
+func (this *ClientMgr) GetStoreKey(contractAddress string, key []byte) ([]byte, error) {
```
Comment (laizy) [inline client/client.go:409]: ditto

File: client/client.go [hunk 3]
```diff
@@ -350,6 +387,51 @@ func (this *ClientMgr) PreExecTransaction(mutTx *types.MutableTransaction) (*sdk
...
+		contractAddr := common.ToArrayReverse(contractAddrBytes)
+		newKey = append(newKey, contractAddr...)
+		newKey = append(newKey, key...)
+	} else {
+		newKey = append(newKey, key...)
+	}
+	return newKey, nil
+}
+
+func (this *ClientMgr) GetStoreProof(key []byte) (*sdkcom.StoreProof, error) {
```
Comment (laizy) [inline client/client.go:423]: ditto

File: common/define.go [hunk 4]
```diff
@@ -278,3 +279,10 @@ type GlobalParam struct {
 	Key   string
 	Value string
 }
+
+type StoreProof struct {
```
Comment (laizy) [inline common/define.go:283]: rename to Layer2StoreProof

File: utils/store_proof.go [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+package utils
+
+import (
+	"io"
+
+	"github.com/ontio/ontology/common"
+	"github.com/tendermint/iavl"
+)
+
+type StoreProof iavl.RangeProof
```
Comment (laizy) [inline utils/store_proof.go:28]: rename to Layer2StoreProof

File: client/rpc.go [hunk 6]
```diff
@@ -179,6 +179,10 @@ func (this *RpcClient) sendRawTransaction(qid string, tx *types.Transaction, isP
 	return this.sendRpcRequest(qid, RPC_SEND_TRANSACTION, params)
 }
 
+func (this *RpcClient) getStoreProof(qid string, key []byte) ([]byte, error) {
```
Comment (laizy) [inline client/rpc.go:182]: rename to getLayer2StoreProof

File: client/rest.go [hunk 7]
```diff
@@ -193,6 +193,11 @@ func (this *RestClient) sendRawTransaction(qid string, tx *types.Transaction, is
 	return this.sendRestPostRequest(common.SerializeToBytes(tx), reqPath, reqValues)
 }
 
+func (this RestClient) getStoreProof(qid string, key []byte) ([]byte, error) {
```
Comment (laizy) [inline client/rest.go:196]: `*RestClient` and  `getLayer2StoreProof`

File: client/define.go [hunk 8]
```diff
@@ -45,6 +45,7 @@ type OntologyClient interface {
 	getMemPoolTxState(qid, txHash string) ([]byte, error)
 	getMemPoolTxCount(qid string) ([]byte, error)
 	sendRawTransaction(qid string, tx *types.Transaction, isPreExec bool) ([]byte, error)
+	getStoreProof(qid string, key []byte) ([]byte, error)
```
Comment (laizy) [inline client/define.go:48]: rename to getLayer2StoreProof

**Review by laizy — CHANGES_REQUESTED** [review]

File: client/client.go (hunk 1 above)
Comment (blockchain-develop) [reply client/client.go]: done

File: client/client.go (hunk 2 above)
Comment (blockchain-develop) [reply client/client.go]: done

File: client/client.go (hunk 3 above)
Comment (blockchain-develop) [reply client/client.go]: done

File: common/define.go (hunk 4 above)
Comment (blockchain-develop) [reply common/define.go]: done

File: utils/store_proof.go (hunk 5 above)
Comment (blockchain-develop) [reply utils/store_proof.go]: done

File: client/rpc.go (hunk 6 above)
Comment (blockchain-develop) [reply client/rpc.go]: done

File: client/rest.go (hunk 7 above)
Comment (blockchain-develop) [reply client/rest.go]: done

File: client/define.go (hunk 8 above)
Comment (blockchain-develop) [reply client/define.go]: done

File: client/rest.go [hunk 9]
```diff
@@ -193,7 +193,7 @@ func (this *RestClient) sendRawTransaction(qid string, tx *types.Transaction, is
 	return this.sendRestPostRequest(common.SerializeToBytes(tx), reqPath, reqValues)
 }
 
-func (this RestClient) getStoreProof(qid string, key []byte) ([]byte, error) {
+func (this RestClient) getLayer2StoreProof(qid string, key []byte) ([]byte, error) {
```
Comment (laizy) [inline client/rest.go]: `(this RestClient)` to `(this *RestClient)`

**Review by laizy — APPROVED** [review]

**Pull request merged**